lib/failure_detector.ml

```diff
@@ -0,0 +1,146 @@
+  acknowledges : (int, unit Lwt_condition.t) Base.Hashtbl.t;
+  mutable sequence_number : int;
+}
+
+(* Helper to increase the round *)
+let next_seq_no t =
+  let sequence_number = t.sequence_number in
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_message message client (recipient : Peer.t) =
+  let message = Util.Encoding.pack bin_writer_message message in
+  Client.send_to client message recipient
+
+let send_ping_to client peer =
+  match peer with
+  | None -> failwith "Trying to send message to nobody)"
+  | Some peer -> send_message Ping client peer
+
+let send_acknowledge_to = send_message Acknowledge
+
+let send_ping_request_to client (recipient : Peer.t) =
+  send_message (PingRequest recipient.address) client recipient
+
+(* TODO: Totally not sure about this...
+   I don't know if I can get rid of it and only use `wait_ack_timeout` instead
+    But I feel like Lwt_condition allows ms to manage async thread and could be useful *)
+let wait_ack t sequence_number =
+  let cond =
+    Base.Hashtbl.find_or_add t.acknowledges sequence_number
+      ~default:Lwt_condition.create in
+  Lwt_condition.wait cond
+
+(* TODO: An async thread should be started here, with a basic sleep, but not sure at all...
+     And we should return Either Ok Timeout *)
+let wait_ack_timeout t sequence_number timeout =
+  Lwt.pick
+    [
+      (let _ = Lwt_unix.sleep (Float.of_int timeout) in
+       Lwt.return @@ Result.Error "Timeout reached");
+      (let _ = wait_ack t sequence_number in
+       Lwt.return @@ Result.Ok "Successfully received acknowledge");
+    ]
+
+(** Basic random shuffle, see https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle*)
+let knuth_shuffle known_peers =
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
+  let initial_array_length = Array.length shuffled_array in
+  for i = initial_array_length - 1 downto 1 do
+    let k = Random.int (i + 1) in
+    let x = shuffled_array.(k) in
+    shuffled_array.(k) <- shuffled_array.(i);
+    shuffled_array.(i) <- x
+  done;
+  Array.to_list shuffled_array
+
+(* Regarding the SWIM protocol, the list of peers is not ordered.
+   Hence, I basically went for a shuffle after adding the new peer *)
+let add_peer (peer : Peer.t) (peer_to_add : Peer.t) =
+  Base.Hashtbl.add peer.peers ~key:peer_to_add.address ~data:peer_to_add
+
+(* Regarding the SWIM protocol, if peer A cannot get ACK from peer B (timeout):
+   A sets B as `suspicious`
+   A randomly picks one (or several, should it also be randomly determined?) peer(s) from its list
+   and ask him/them to ping B.*)
+(* This function return the random peer, to which we will ask to ping the first peer *)
+let rec pick_random_peer_addresses (peers : (Address.t, Peer.t) Base.Hashtbl.t)
+    number_of_peers =
+  let addresses = knuth_shuffle @@ Base.Hashtbl.keys peers in
+  match addresses with
+  | [] -> failwith "pick_random_peers"
+  | elem :: _ ->
```

Comment:
If you'd like, the OCaml idiom here is "piping"

```ocaml
let addresses =
  peers
  |> Base.Hashtbl.keys
  |> knuth_shuffle
```

We tend to enjoy that this makes the flow of data through transformations very obvious 😄. OCaml format might reduce this to one line, but that's okay too.

Reply:
Thanks! Don't have (yet) the habit to use `|>`!